### PR TITLE
fix: asset fetch failure retry with coil

### PR DESCRIPTION
**Pull request by trOnk12**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
    - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
    - [ ] contains a reference JIRA issue number like `SQPIT-764`
    - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
    - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

Sometimes the avatar byte code seems to be corrupted I could not find the cause of the corruption possibly due to cancellation not correctly coorporating, but it is a wild guess. 

### Solutions

When coil returns a failure, we reexecute AssetFetcher with a retry attempt and delete the corrupted asset to redownload it.

### Dependencies (Optional)

_If there are some other pull requests related to this one (e.g. new releases of frameworks), specify them here._

Needs releases with:

https://github.com/wireapp/kalium/pull/1357

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

#### How to Test

_Briefly describe how this change was tested and if applicable the exact steps taken to verify that it works as expected._

### Notes (Optional)

_Specify here any other facts that you think are important for this issue._

### Attachments (Optional)

_Attachments like images, videos, etc. (drag and drop in the text box)_ 

----
#### PR Post Submission Checklist for internal contributors (Optional)

- [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

- [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by AndroidBob** [comment]
Build [2](https://10.10.124.134/job/wire-android-reloaded/job/PR-1345/2/) **failed**.

**Comment by AndroidBob** [comment]
Build [4](https://10.10.124.134/job/wire-android-reloaded/job/PR-1345/4/) **failed**.

**Comment by AndroidBob** [comment]
Build [5](https://10.10.124.134/job/wire-android-reloaded/job/PR-1345/5/) **failed**.

**Comment by AndroidBob** [comment]
Build [6](https://10.10.124.134/job/wire-android-reloaded/job/PR-1345/6/) **failed**.

**Comment by AndroidBob** [comment]
Build [7](https://10.10.124.134/job/wire-android-reloaded/job/PR-1345/7/) **failed**.

**Comment by AndroidBob** [comment]
Build [9](https://10.10.124.134/job/wire-android-reloaded/job/PR-1345/9/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(5552300)(5552300)-dev-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/compat/com.wire.android-vv0.0.1(5552300)(5552300)-dev-compat.apk)

**Comment by AndroidBob** [comment]
Build [11](https://10.10.124.134/job/wire-android-reloaded/job/PR-1345/11/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(5552785)(5552785)-dev-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/compat/com.wire.android-vv0.0.1(5552785)(5552785)-dev-compat.apk)

**Review by alexandreferris — APPROVED** [review]

File: app/src/main/kotlin/com/wire/android/util/ui/AssetImageFetcher.kt [hunk 1]
```diff
@@ -44,14 +58,26 @@ internal class AssetImageFetcher(
     class Factory(
         private val getPublicAssetUseCase: GetAvatarAssetUseCase,
         private val getPrivateAssetUseCase: GetMessageAssetUseCase,
-        private val resources: Resources,
+        private val deleteAssetUseCase: DeleteAssetUseCase,
```
Comment (yamilmedina) [inline app/src/main/kotlin/com/wire/android/util/ui/AssetImageFetcher.kt:61]:
```suggestion
        private val deleteAsset: DeleteAssetUseCase,
```

File: app/src/main/kotlin/com/wire/android/util/ui/WireSessionImageLoader.kt [hunk 2]
```diff
@@ -23,41 +29,72 @@ import com.wire.kalium.logic.feature.asset.GetMessageAssetUseCase
...
+        const val RETRY_INCREMENT_ATTEMPT_PER_STEP = 1
+    }
 
     /**
      * Attempts to paint an Image using [asset], falling back to [fallbackData] if [asset] is null.
      * Just like [rememberAsyncImagePainter], [fallbackData] can be anything that [Coil] accepts.
+     *
+     * currently Coil does not have a friendly API to retry a failing image request, so we have to do it ourselves.
+     * adding retry_hash is a workaround to force Coil to retry the request.
+     * see https://github.com/coil-kt/coil/issues/884
```
Comment (yamilmedina) [inline app/src/main/kotlin/com/wire/android/util/ui/WireSessionImageLoader.kt:42]: Nice workaround 🙌

**Review by yamilmedina — APPROVED** [review]
Just a small naming convention, fantastic solution ! 🪁

**Pull request merged**